### DISCUSSION TITLE
How can I implement the backward pass while providing a forward pass implementation for all backends?

### DISCUSSION THREAD

**Discussion by nazo6** [tracel-ai/burn#4535] [Q&A] [answered]
Hello, I'm trying to implement a custom backward pass like `torch.autograd.Function` and came across [this](https://burn.dev/books/burn/advanced/backend-extension/index.html) documentation.

However, implementing the backward pass in Autodiff this way means it's impossible to provide a default implementation of the forward pass for all backends.
In other words, the following code will result in an error because specialization is not implemented in Rust.
```rust
pub trait MyCustomBackend: Backend {
    fn some_func() {
        ... default forward pass impl
    }
}

impl<B: Backend> MyCustomBackend for B {}
impl<B: MyCustomBackend, C: CheckpointStrategy> SgBackend for Autodiff<B, C> { // <- conflicting implementations of trait
    fn some_func() {
        ... backward pass impl
    }
}
```

Of course, it is possible to implement it by listing all possible backends as shown below, but I don't think that is desirable.
```rust
impl MyCustomBackend for CubeBackend {}
impl MyCustomBackend for NdArray {}
impl MyCustomBackend for Candle {}
```

Is there a way to provide a backward pass implementation for Autodiff while still allowing other backends to use the default forward pass implementation?

**Comment by laggui** [comment, marked as answer]
Hello! That's right, the backend extension approach is currently the intended path. But I understand that it's quite verbose for a single op, mainly due to lack of specialization in Rust.

This could potentially be improved if we provide a function-like API for custom ops more specifically, not a whole backend extension. It would allow users to define the forward/backward logic once and let Burn handle the dispatching.

**Reply by nazo6** [reply]
Thank you for your answer. Yes, I think that having such an API would make implementation much easier.

**Comment by nazo6** [comment]
BTW, I have found that "detach trick" works for my use case.
https://ajkdrag.in/2-Zettels/detach-trick-for-straight-through-estimators

In the operation below, a Heaviside step function is used for the forward pass and a sigmoid function for the gradient calculation.
```rust
use burn::prelude::Backend;
use burn::tensor::Tensor;
use burn::tensor::activation::sigmoid;

pub fn heaviside_surrogate_sigmoid<B: Backend, const D: usize>(x: Tensor<B, D>) -> Tensor<B, D> {
    let y_fwd = x.clone().greater_equal_elem(0.0).float();

    let y_bwd = sigmoid(x);

    (y_fwd - y_bwd.clone().detach()).detach() + y_bwd
}

#[test]
fn test() {
    use burn::backend::{Autodiff, NdArray};

    type B = Autodiff<NdArray<f32>>;
    let device = Default::default();

    let x = Tensor::<B, 1>::from_floats([0.5], &device).require_grad();

    let y = heaviside_surrogate_sigmoid(x.clone());

    // Forward pass
    let y_val = y.clone().into_scalar();
    println!("[Forward pass] y_val: {}", y_val);
    assert_eq!(y_val, 1.0, "Forward pass should use Heaviside");

    // Backward pass
    let mut grads = y.backward();
    let grad_x = x.grad_remove(&mut grads).expect("Should have gradient");
    let grad_val = grad_x.into_scalar();

    let s = 1.0 / (1.0 + (-0.5f32).exp());
    let expected_grad = s * (1.0 - s);

    println!(
        "[Backward pass] grad_val: {}, expected_grad: {}",
        grad_val, expected_grad
    );
    assert!(
        (grad_val - expected_grad).abs() < 1e-6,
        "Backward pass should use Sigmoid gradient. Expected {}, got {}",
        expected_grad,
        grad_val
    );
}
```
